runtime/compiler/env/VMJ9Server.cpp

```diff
@@ -692,9 +692,48 @@ TR_J9ServerVM::getStaticReferenceFieldAtAddress(uintptrj_t fieldAddress)
 bool
 TR_J9ServerVM::stackWalkerMaySkipFrames(TR_OpaqueMethodBlock *method, TR_OpaqueClassBlock *clazz)
    {
+   if(!method)
+      {
+      return false;
+      }
+
    JITServer::ServerStream *stream = _compInfoPT->getMethodBeingCompiled()->_stream;
-   stream->write(JITServer::MessageType::VM_stackWalkerMaySkipFrames, method, clazz);
-   return std::get<0>(stream->read<bool>());
+   auto *vmInfo = _compInfoPT->getClientData()->getOrCacheVMInfo(stream);
+   
+   if (vmInfo->_jlrMethodInvoke == NULL)
+      {
+      // Cached information could be outdated. Ask the client again.
+      stream->write(JITServer::MessageType::VM_stackWalkerMaySkipFrames, JITServer::Void());
+      vmInfo->_jlrMethodInvoke = std::get<0>(stream->read<J9Method*, TR_OpaqueClassBlock*>());
+      if (vmInfo->_jlrMethodInvoke == NULL)
+         return true;
+      }
+   if (vmInfo->_jlrMethodInvoke == ((J9Method *) method))
+      {
+      return true;
+      }
+   if(!clazz)
+      {
+      return false;
+      }
+#if defined(J9VM_OPT_SIDECAR)
+   if (vmInfo->_srMethodAccessorClass == NULL)
+      {
+      // Cached information could be outdated. Ask the client again.
+      stream->write(JITServer::MessageType::VM_stackWalkerMaySkipFrames, JITServer::Void());
+      vmInfo->_srMethodAccessorClass = std::get<1>(stream->read<J9Method*, TR_OpaqueClassBlock*>());
+      }
+   if (vmInfo->_srMethodAccessorClass != NULL && TR_J9ServerVM::isInstanceOf( clazz, vmInfo->_srMethodAccessorClass ,false))
+      {
+      return true;
+      }
+   if (vmInfo->_srConstructorAccessorClass != NULL && TR_J9ServerVM::isInstanceOf( clazz, vmInfo->_srConstructorAccessorClass ,false))
```

Comment:
I see that the VM sets `srConstructorAccessor` and `srMethodAccessor` in succession, so if one can be null the other one can be too.
I also suggest getting all 3 fields that we need from the client in one message, rather than sending separate messages.

Reply:
Will do that. The question is where should we send this message? 

I could populate [this if block](https://github.com/eclipse/openj9/pull/8680/files/0fe41d824311d3ad383b07d5dac7e4e919910602#diff-770fc7473e17fd36db57f82532b90d82R703) (line 703). Add `srConstructorAccessor == NULL || srMethodAccessor  == NULL` in if condition, query and cache three values at the same time. 

Also I noticed that `srConstructorAccessor` and `srMethodAccessor` are protected by `#if defined(J9VM_OPT_SIDECAR)` in `TR_J9VM::stackWalkerMaySkipFrames` implementation, however they are not protected in `J9JavaVM` definition. I am not sure what `J9VM_OPT_SIDECAR` is exactly doing, so I hope my suggestion won't break anything in the case of `J9VM_OPT_SIDECAR` is not defined.

Reply:
I'd say if the methods that use `srConstructorAccessor` or `srMethodAccessor` in the VM code are not protected by `J9VM_OPT_SIDECAR` we should remove that protection as well.

> The question is where should we send this message?

Your implementation is ok, but it can be optimized because sometimes even with partial information we can answer the frontend query. So I am thinking something along these lines:
```
get cached information
bool freshInfo = false;
if (vmInfo->_jlrMethodInvoke == NULL)
   send message and get all 3 pieces that we want
   store them into vmInfo
   freshInfo = true;
   if (vmInfo->_jlrMethodInvoke == NULL)
        return true;
if (vmInfo->_jlrMethodInvoke == ((J9Method *) method))
  return true;
if (!clazz)
   return false;
if (vmInfo->_srMethodAccessorClass == NULL && !freshInfo)
     send message and get all 3 pieces that we want
     store them into vmInfo
     freshInfo = true;
  if (vmInfo->_srMethodAccessorClass != NULL && TR_J9ServerVM::isInstanceOf( clazz, vmInfo->_srMethodAccessorClass ,false))
      return true;
repeat for srConstructorAccessor
```

Reply:
Looked into the pseudo code, we actually have two duplicated if checks, so I just kept the one outside. 

``` c++
if (vmInfo->_srMethodAccessorClass == NULL && !freshInfo)
     send message and get all 3 pieces that we want
     store them into vmInfo
     freshInfo = true;
// HERE, this one is removed in latest commit
  if (vmInfo->_srMethodAccessorClass != NULL && TR_J9ServerVM::isInstanceOf( clazz, vmInfo->_srMethodAccessorClass ,false))
      return true;
// HERE
if (vmInfo->_srMethodAccessorClass != NULL && TR_J9ServerVM::isInstanceOf( clazz, vmInfo->_srMethodAccessorClass ,false))
      return true;
```